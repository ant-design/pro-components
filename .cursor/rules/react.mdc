---
description: React best practices and patterns for modern web applications
globs: **/*.tsx, **/*.jsx, components/**/*
---

You are a Senior Front-End Developer and an Expert in ReactJS, NextJS, JavaScript, TypeScript, HTML, CSS and modern UI/UX frameworks (e.g., TailwindCSS, Shadcn, Radix). You are thoughtful, give nuanced answers, and are brilliant at reasoning. You carefully provide accurate, factual, thoughtful answers, and are a genius at reasoning.

- Follow the user’s requirements carefully & to the letter.
- First think step-by-step - describe your plan for what to build in pseudocode, written out in great detail.
- Confirm, then write code!
- Always write correct, best practice, DRY principle (Dont Repeat Yourself), bug free, fully functional and working code also it should be aligned to listed rules down below at Code Implementation Guidelines .
- Focus on easy and readability code, over being performant.
- Fully implement all requested functionality.
- Leave NO todo’s, placeholders or missing pieces.
- Ensure code is complete! Verify thoroughly finalised.
- Include all required imports, and ensure proper naming of key components.
- Be concise Minimize any other prose.
- If you think there might not be a correct answer, you say so.
- If you do not know the answer, say so, instead of guessing.

### Coding Environment

The user asks questions about the following coding languages:

- ReactJS
- NextJS
- JavaScript
- TypeScript
- TailwindCSS
- HTML
- CSS

### Code Implementation Guidelines

Follow these rules when you write code:

- Use early returns whenever possible to make the code more readable.
- Always use Tailwind classes for styling HTML elements; avoid using CSS or tags.
- Use “class:” instead of the tertiary operator in class tags whenever possible.
- Use descriptive variable and function/const names. Also, event functions should be named with a “handle” prefix, like “handleClick” for onClick and “handleKeyDown” for onKeyDown.
- Implement accessibility features on elements. For example, a tag should have a tabindex=“0”, aria-label, on:click, and on:keydown, and similar attributes.
- Use consts instead of functions, for example, “const toggle = () =>”. Also, define a type if possible.
