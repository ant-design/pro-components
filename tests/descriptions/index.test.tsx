import ProDescriptions from '@ant-design/pro-descriptions';
import type { ProCoreActionType } from '@ant-design/pro-utils';
import { cleanup, render, waitFor } from '@testing-library/react';
import { Button, Input } from 'antd';
import React, { act } from 'react';
afterEach(() => {
  cleanup();
});

describe('descriptions', () => {
  afterEach(() => {
    cleanup();
  });
  it('🥩 descriptions render valueEnum when data = 0', async () => {
    const { container } = render(
      <ProDescriptions
        columns={[
          {
            dataIndex: 'status',
            title: '状态',
            valueEnum: {
              0: { text: '关闭', status: 'Default' },
              1: { text: '运行中', status: 'Processing' },
              2: { text: '已上线', status: 'Success' },
              3: { text: '异常', status: 'Error' },
            },
          },
        ]}
        request={async () => ({
          data: {
            status: 0,
          },
        })}
      />,
    );

    await waitFor(() =>
      expect(
        container.querySelector('span.ant-badge-status-text')?.innerHTML,
      ).toBe('关闭'),
    );
  });

  it('🎏 onLoadingChange test', async () => {
    const fn = vi.fn();
    render(
      <ProDescriptions
        size="small"
        onLoadingChange={fn}
        columns={[
          {
            dataIndex: 'money',
            valueType: 'money',
          },
        ]}
        request={async () => {
          return {
            data: [],
          };
        }}
      />,
    );

    await waitFor(() => {
      expect(fn).toBeCalled();
    });
  });

  it('🎏 loading test', async () => {
    vi.useFakeTimers();
    const html = render(
      <ProDescriptions
        columns={[
          {
            title: 'money',
            dataIndex: 'money',
            valueType: 'money',
          },
        ]}
        request={async () => {
          return new Promise((resolve) => {
            setTimeout(() => {
              resolve({ data: [] });
            }, 5000);
          });
        }}
      />,
    );

    act(() => {
      vi.advanceTimersByTime(2000);
    });

    await waitFor(() => {
      expect(!!html.baseElement.querySelector('.ant-skeleton')).toBeTruthy();
    });

    act(() => {
      html.rerender(
        <ProDescriptions
          columns={[
            {
              title: 'money',
              dataIndex: 'money',
              valueType: 'money',
            },
          ]}
          loading={false}
          request={async () => {
            return new Promise((resolve) => {
              setTimeout(() => {
                resolve({ data: [] });
              }, 5000);
            });
          }}
        />,
      );
    });

    act(() => {
      vi.advanceTimersByTime(2000);
    });

    await waitFor(() => {
      // props 指定为 false 后，无论 request 完成与否都不会出现 spin
      expect(!!html.baseElement.querySelector('.ant-skeleton')).toBeFalsy();
    });

    vi.useRealTimers();
  });

  it('🥩 test reload', async () => {
    const fn = vi.fn();
    vi.useFakeTimers();
    const actionRef = React.createRef<ProCoreActionType>();
    const Reload = () => {
      return (
        <ProDescriptions
          actionRef={actionRef}
          title="高级定义列表 request"
          request={async () => {
            fn();
            return new Promise((resolve) => {
              setTimeout(() => {
                resolve({
                  success: true,
                  data: {
                    id: '这是一段文本',
                    date: '20200730',
                    money: '12121',
                  },
                });
              }, 2000);
            });
          }}
          extra={
            <Button
              type="link"
              id="reload"
              onClick={() => {
                actionRef.current?.reload();
              }}
            >
              刷新
            </Button>
          }
        >
          test reload
          <ProDescriptions.Item label="文本" dataIndex="id" />
          <ProDescriptions.Item
            dataIndex="date"
            label="日期"
            valueType="date"
          />
          <ProDescriptions.Item
            label="money"
            dataIndex="money"
            valueType="money"
            renderFormItem={() => <Input />}
          />
        </ProDescriptions>
      );
    };
    const html = render(<Reload />);

    await act(() => {
      return vi.runOnlyPendingTimers();
    });

    await html.findAllByText('这是一段文本');
    await waitFor(() => {
      expect(fn).toBeCalledTimes(1);
    });
    act(() => {
      html.queryByText('刷新')?.click();
    });
    act(() => {
      actionRef.current?.reload();
    });
    act(() => {
      actionRef.current?.reload();
    });

    await waitFor(() => {
      // 因为有 loading 的控制，所有只会触发两次
      expect(fn).toBeCalledTimes(2);
    });

    vi.useRealTimers();
  });

  it('🥩 test reload by params', async () => {
    const fn = vi.fn();
    vi.useFakeTimers();
    const html = render(
      <ProDescriptions
        title="高级定义列表 request"
        request={async () => {
          fn();
          return Promise.resolve({
            success: true,
            data: { id: '这是一段文本', date: '20200730', money: '12121' },
          });
        }}
        extra={
          <Button type="link" id="reload">
            修改
          </Button>
        }
      >
        <ProDescriptions.Item label="文本" dataIndex="id" />
        <ProDescriptions.Item dataIndex="date" label="日期" valueType="date" />
        <ProDescriptions.Item
          label="money"
          dataIndex="money"
          valueType="money"
        />
      </ProDescriptions>,
    );

    await html.findAllByText('这是一段文本');

    act(() => {
      vi.runOnlyPendingTimers();
    });

    await waitFor(() => {
      expect(fn).toBeCalledTimes(1);
    });

    act(() => {
      html.rerender(
        <ProDescriptions
          title="高级定义列表 request"
          request={async () => {
            fn();
            return Promise.resolve({
              success: true,
              data: { id: '这是一段文本', date: '20200730', money: '12121' },
            });
          }}
          extra={
            <Button type="link" id="reload">
              修改
            </Button>
          }
          params={{ name: 'qixian' }}
        >
          <ProDescriptions.Item label="文本" dataIndex="id" />
          <ProDescriptions.Item
            dataIndex="date"
            label="日期"
            valueType="date"
          />
          <ProDescriptions.Item
            label="money"
            dataIndex="money"
            valueType="money"
          />
        </ProDescriptions>,
      );
    });

    act(() => {
      vi.runOnlyPendingTimers();
    });

    await html.findAllByText('这是一段文本');

    await waitFor(() => {
      expect(fn).toBeCalledTimes(2);
    });

    vi.useRealTimers();
  });

  it('🥩 test request error', async () => {
    const fn = vi.fn();

    render(
      <ProDescriptions
        title="高级定义列表 request"
        request={async () => {
          throw new Error('load error');
        }}
        onRequestError={fn}
        extra={
          <Button type="link" id="reload">
            修改
          </Button>
        }
      >
        <ProDescriptions.Item label="文本" dataIndex="id" />
        <ProDescriptions.Item dataIndex="date" label="日期" valueType="date" />
        <ProDescriptions.Item
          label="money"
          dataIndex="money"
          valueType="money"
        />
      </ProDescriptions>,
    );

    await waitFor(() => {
      expect(fn).toBeCalledTimes(1);
    });
  });

  it('🏊 Progress', async () => {
    const html = render(
      <ProDescriptions>
        <ProDescriptions.Item label="进度条1" valueType="progress">
          40
        </ProDescriptions.Item>
        <ProDescriptions.Item label="进度条2" valueType="progress">
          -1
        </ProDescriptions.Item>
        <ProDescriptions.Item label="进度条3" valueType="progress">
          100
        </ProDescriptions.Item>
      </ProDescriptions>,
    );
    await waitFor(() => {
      expect(
        html.baseElement.querySelector('.ant-progress-text')?.textContent,
      ).toEqual('40%');
    });

    await waitFor(() => {
      expect(
        !!html.baseElement
          .querySelectorAll('.ant-progress-text')?.[1]
          ?.querySelector('.anticon-close-circle'),
      ).toBeTruthy();
      expect(
        !!html.baseElement
          .querySelectorAll('.ant-progress-text')?.[2]
          ?.querySelector('.anticon-check-circle'),
      ).toBeTruthy();
    });
  });

  it('🏊 ProDescriptions support order', async () => {
    const html = render(
      <ProDescriptions
        dataSource={{
          title: 'test',
        }}
        columns={[
          {
            title: '标题',
            dataIndex: 'title',
            valueType: 'text',
            order: 100,
          },
        ]}
      >
        <ProDescriptions.Item order={9} label="进度条1" valueType="progress">
          40
        </ProDescriptions.Item>
        <ProDescriptions.Item label="进度条2" valueType="progress">
          -1
        </ProDescriptions.Item>
        <ProDescriptions.Item order={8} label="进度条3" valueType="progress">
          100
        </ProDescriptions.Item>
      </ProDescriptions>,
    );
    expect(html.asFragment()).toMatchSnapshot();
  });

  it('📝 typography support and copy', async () => {
    const wrapper = render(
      <ProDescriptions
        title="dataSource and columns"
        dataSource={{
          id: '这是一段文本columns',
          date: '20200809',
          money: '1212100',
          state: 'all',
          state2: 'open',
        }}
        columns={[
          {
            title: '文本',
            key: 'text',
            dataIndex: 'id',
            ellipsis: true,
            copyable: true,
          },
        ]}
      />,
    );

    await waitFor(() => {
      expect(
        wrapper.baseElement.querySelector(
          'span.ant-descriptions-item-content button.ant-typography-copy',
        ),
      ).toBeTruthy();
    });

    wrapper.rerender(
      <ProDescriptions
        title="dataSource and columns"
        dataSource={{
          id: '这是一段文本columns',
          date: '20200809',
          money: '1212100',
          state: 'all',
          state2: 'open',
        }}
        columns={[
          {
            title: '文本',
            key: 'text',
            dataIndex: 'id',
          },
        ]}
      />,
    );

    await waitFor(() => {
      expect(
        wrapper.baseElement.querySelectorAll(
          '.ant-descriptions-item-content .ant-typography-copy',
        ).length,
      ).toBe(0);
    });

    wrapper.unmount();
  });
});
