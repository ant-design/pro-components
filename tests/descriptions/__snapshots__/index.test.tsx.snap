// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`descriptions > üèä ProDescriptions support order 1`] = `
<DocumentFragment>
  <div
    class="ant-descriptions ant-pro-descriptions"
  >
    <div
      class="ant-descriptions-view"
    >
      <table>
        <tbody>
          <tr
            class="ant-descriptions-row"
          >
            <td
              class="ant-descriptions-item"
              colspan="1"
            >
              <div
                class="ant-descriptions-item-container"
              >
                <span
                  class="ant-descriptions-item-label"
                >
                  Ê†áÈ¢ò
                </span>
                <span
                  class="ant-descriptions-item-content"
                  style="min-width: 0;"
                >
                  test
                </span>
              </div>
            </td>
            <td
              class="ant-descriptions-item"
              colspan="1"
            >
              <div
                class="ant-descriptions-item-container"
              >
                <span
                  class="ant-descriptions-item-label"
                >
                  ËøõÂ∫¶Êù°1
                </span>
                <span
                  class="ant-descriptions-item-content"
                  style="min-width: 0;"
                >
                  <div
                    aria-valuemax="100"
                    aria-valuemin="0"
                    aria-valuenow="40"
                    class="ant-progress ant-progress-status-active ant-progress-line ant-progress-line-align-end ant-progress-line-position-outer ant-progress-show-info ant-progress-small"
                    role="progressbar"
                    style="min-width: 100px; max-width: 320px;"
                  >
                    <div
                      class="ant-progress-outer"
                      style="width: 100%;"
                    >
                      <div
                        class="ant-progress-inner"
                      >
                        <div
                          class="ant-progress-bg ant-progress-bg-outer"
                          style="width: 40%; height: 6px; --progress-percent: 0.4;"
                        />
                      </div>
                      <span
                        class="ant-progress-text ant-progress-text-end ant-progress-text-outer"
                        title="40%"
                      >
                        40%
                      </span>
                    </div>
                  </div>
                </span>
              </div>
            </td>
            <td
              class="ant-descriptions-item"
              colspan="1"
            >
              <div
                class="ant-descriptions-item-container"
              >
                <span
                  class="ant-descriptions-item-label"
                >
                  ËøõÂ∫¶Êù°3
                </span>
                <span
                  class="ant-descriptions-item-content"
                  style="min-width: 0;"
                >
                  <div
                    aria-valuemax="100"
                    aria-valuemin="0"
                    aria-valuenow="100"
                    class="ant-progress ant-progress-status-success ant-progress-line ant-progress-line-align-end ant-progress-line-position-outer ant-progress-show-info ant-progress-small"
                    role="progressbar"
                    style="min-width: 100px; max-width: 320px;"
                  >
                    <div
                      class="ant-progress-outer"
                      style="width: 100%;"
                    >
                      <div
                        class="ant-progress-inner"
                      >
                        <div
                          class="ant-progress-bg ant-progress-bg-outer"
                          style="width: 100%; height: 6px; --progress-percent: 1;"
                        />
                      </div>
                      <span
                        class="ant-progress-text ant-progress-text-end ant-progress-text-outer"
                      >
                        <span
                          aria-label="check-circle"
                          class="anticon anticon-check-circle"
                          role="img"
                        >
                          <svg
                            aria-hidden="true"
                            data-icon="check-circle"
                            fill="currentColor"
                            focusable="false"
                            height="1em"
                            viewBox="64 64 896 896"
                            width="1em"
                          >
                            <path
                              d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm193.5 301.7l-210.6 292a31.8 31.8 0 01-51.7 0L318.5 484.9c-3.8-5.3 0-12.7 6.5-12.7h46.9c10.2 0 19.9 4.9 25.9 13.3l71.2 98.8 157.2-218c6-8.3 15.6-13.3 25.9-13.3H699c6.5 0 10.3 7.4 6.5 12.7z"
                            />
                          </svg>
                        </span>
                      </span>
                    </div>
                  </div>
                </span>
              </div>
            </td>
          </tr>
          <tr
            class="ant-descriptions-row"
          >
            <td
              class="ant-descriptions-item"
              colspan="3"
            >
              <div
                class="ant-descriptions-item-container"
              >
                <span
                  class="ant-descriptions-item-label"
                >
                  ËøõÂ∫¶Êù°2
                </span>
                <span
                  class="ant-descriptions-item-content"
                  style="min-width: 0;"
                >
                  <div
                    aria-valuemax="100"
                    aria-valuemin="0"
                    aria-valuenow="-1"
                    class="ant-progress ant-progress-status-exception ant-progress-line ant-progress-line-align-end ant-progress-line-position-outer ant-progress-show-info ant-progress-small"
                    role="progressbar"
                    style="min-width: 100px; max-width: 320px;"
                  >
                    <div
                      class="ant-progress-outer"
                      style="width: 100%;"
                    >
                      <div
                        class="ant-progress-inner"
                      >
                        <div
                          class="ant-progress-bg ant-progress-bg-outer"
                          style="width: 0%; height: 6px; --progress-percent: 0;"
                        />
                      </div>
                      <span
                        class="ant-progress-text ant-progress-text-end ant-progress-text-outer"
                      >
                        <span
                          aria-label="close-circle"
                          class="anticon anticon-close-circle"
                          role="img"
                        >
                          <svg
                            aria-hidden="true"
                            data-icon="close-circle"
                            fill="currentColor"
                            fill-rule="evenodd"
                            focusable="false"
                            height="1em"
                            viewBox="64 64 896 896"
                            width="1em"
                          >
                            <path
                              d="M512 64c247.4 0 448 200.6 448 448S759.4 960 512 960 64 759.4 64 512 264.6 64 512 64zm127.98 274.82h-.04l-.08.06L512 466.75 384.14 338.88c-.04-.05-.06-.06-.08-.06a.12.12 0 00-.07 0c-.03 0-.05.01-.09.05l-45.02 45.02a.2.2 0 00-.05.09.12.12 0 000 .07v.02a.27.27 0 00.06.06L466.75 512 338.88 639.86c-.05.04-.06.06-.06.08a.12.12 0 000 .07c0 .03.01.05.05.09l45.02 45.02a.2.2 0 00.09.05.12.12 0 00.07 0c.02 0 .04-.01.08-.05L512 557.25l127.86 127.87c.04.04.06.05.08.05a.12.12 0 00.07 0c.03 0 .05-.01.09-.05l45.02-45.02a.2.2 0 00.05-.09.12.12 0 000-.07v-.02a.27.27 0 00-.05-.06L557.25 512l127.87-127.86c.04-.04.05-.06.05-.08a.12.12 0 000-.07c0-.03-.01-.05-.05-.09l-45.02-45.02a.2.2 0 00-.09-.05.12.12 0 00-.07 0z"
                            />
                          </svg>
                        </span>
                      </span>
                    </div>
                  </div>
                </span>
              </div>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</DocumentFragment>
`;
